compiler/rustc_codegen_ssa/src/mir/mod.rs

```diff
@@ -212,24 +213,19 @@ pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
 
     fx.per_local_var_debug_info = fx.compute_per_local_var_debug_info(&mut start_bx);
 
-    // Evaluate all required consts; codegen later assumes that CTFE will never fail.
-    let mut all_consts_ok = true;
-    for const_ in &mir.required_consts {
-        if let Err(err) = fx.eval_mir_constant(const_) {
-            all_consts_ok = false;
-            match err {
-                // errored or at least linted
-                ErrorHandled::Reported(_) => {}
-                ErrorHandled::TooGeneric => {
-                    span_bug!(const_.span, "codegen encountered polymorphic constant: {:?}", err)
-                }
-            }
+    // Rust post-monomorphization checks; we later rely on them.
+    match mir.post_mono_checks(cx.tcx(), ty::ParamEnv::reveal_all(), |c| Ok(fx.monomorphize(c))) {
+        Ok(()) => {}
+        Err(ErrorHandled::TooGeneric(span)) => {
+            cx.tcx().sess.diagnostic().emit_bug(errors::PolymorphicConstantTooGeneric { span });
+        }
+        Err(ErrorHandled::Reported(_, span)) => {
+            cx.tcx().sess.emit_err(errors::ErroneousConstant { span });
```

Comment:
Maybe handle TooGeneric entirely inside post_mono_checks, and for Reported emit the error inside it and return `Err(())` to allow the backend to skip codegen for the function?

Reply:
One of the goals was to share this code with const-eval, and in const-eval TooGeneric can unfortunately occur sine const generics will (try to) evaluate generic MIR. At least I got tons of ICEs when I tried to handle TooGeneric in post_mono_checks. (There's a doc comment at post_mono_checks explaining this.)